backend/django/core/views/api_annotate.py

```diff
@@ -437,20 +444,67 @@ def data_unlabeled_table(request, project_pk):
     for d in unlabeled_data:
         serialized_data = DataSerializer(d, many=False).data
 
-        sorted_similarity_pairs = []
-        for pair in serialized_data["similarityPair"]:
-            sorted_similarity_pairs.append(pair)
         temp = {
             "Text": serialized_data["text"],
             "metadata": serialized_data["metadata"],
-            "similarityPair": serialized_data["similarityPair"],
             "ID": d.id,
         }
         data.append(temp)
 
     return Response({"data": data})
 
 
+@api_view(["POST"])
+@permission_classes([AllowAny])
+def embeddings_calculations(request):
+    """This calculates embeddings for a given array of strings.
+
+    Args:
+        strings: The array of strings
+        request: The POST request
+    Returns:
+        data: a list of data information
+    """
+
+    embeddings = embeddings_model.encode(request.data["strings"])
+
+    return Response(embeddings)
+
+
+@api_view(["GET"])
+@permission_classes((IsAdminOrCreator,))
+def embeddings_comparison(request, project_pk):
+    """This finds the highest scoring labels when comparing cosine similarity scores of
+    their embeddingsfor a given input string.
+
+    Args:
+        text: The input string
+        request: The GET request
+        project_pk: Primary key of the project
+    Returns:
+        data: a list of data information
+    """
+    project = Project.objects.get(pk=project_pk)
+    project_labels = Label.objects.filter(project=project)
+    project_labels_embeddings = list(
+        LabelEmbeddings.objects.filter(label_id__in=project_labels).values_list(
+            "embedding", flat=True
+        )
+    )
+
+    text = request.GET.get("text")
+    text_embedding = embeddings_model.encode(text)
+
+    cosine_scores = util.pytorch_cos_sim(text_embedding, project_labels_embeddings)
+    values, indices = cosine_scores[0].topk(5)
+
+    suggestions = []
+    for index in indices:
+        suggestions.append(LabelSerializer(project_labels[int(index)]).data)
```

Comment:
So, I'm running into an interesting error in this line. It only starts showing up after new data is added to an existing project. 

Basically, when I first make a project with many labels, everything works as expected. However, if I add new data to the project using the update page, and then try to go to the annotate page, I get the following:

```
[02/Feb/2022 22:29:59] "GET /api/comparisons/5?text=Agroecology%20and%20Sustainable%20Agriculture HTTP/1.1" 301 0
backend              | Internal Server Error: /api/comparisons/5/
backend              | Traceback (most recent call last):
backend              |   File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
backend              |     response = get_response(request)
backend              |   File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
backend              |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
backend              |   File "/usr/local/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
backend              |     return view_func(*args, **kwargs)
backend              |   File "/usr/local/lib/python3.8/site-packages/django/views/generic/base.py", line 70, in view
backend              |     return self.dispatch(request, *args, **kwargs)
backend              |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
backend              |     response = self.handle_exception(exc)
backend              |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
backend              |     self.raise_uncaught_exception(exc)
backend              |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
backend              |     raise exc
backend              |   File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
backend              |     response = handler(request, *args, **kwargs)
backend              |   File "/usr/local/lib/python3.8/site-packages/rest_framework/decorators.py", line 50, in handler
backend              |     return func(*args, **kwargs)
backend              |   File "/code/core/views/api_annotate.py", line 503, in embeddings_comparison
backend              |     suggestions.append(LabelSerializer(project_labels[int(index)]).data)
backend              |   File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 318, in __getitem__
backend              |     return qs._result_cache[0]
backend              | IndexError: list index out of range
```

Reply:
Still getting this error. Did you have trouble recreating it?

Reply:
If the data you add in project update has overlap with the data already in the project, it tries to make the embeddings for the duplicate data and then throws this error

```
dev-celery_worker-1  | The above exception was the direct cause of the following exception:
dev-celery_worker-1  | 
dev-celery_worker-1  | Traceback (most recent call last):
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/celery/app/trace.py", line 451, in trace_task
dev-celery_worker-1  |     R = retval = fun(*args, **kwargs)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/celery/app/trace.py", line 734, in __protected_call__
dev-celery_worker-1  |     return self.run(*args, **kwargs)
dev-celery_worker-1  |   File "/code/core/tasks.py", line 67, in send_label_embeddings_task
dev-celery_worker-1  |     generate_label_embeddings(project_pk)
dev-celery_worker-1  |   File "/code/core/utils/util.py", line 257, in generate_label_embeddings
dev-celery_worker-1  |     LabelEmbeddings.objects.bulk_create(label_embeddings)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
dev-celery_worker-1  |     return getattr(self.get_queryset(), name)(*args, **kwargs)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 514, in bulk_create
dev-celery_worker-1  |     returned_columns = self._batched_insert(
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 1287, in _batched_insert
dev-celery_worker-1  |     inserted_rows.extend(self._insert(
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 1270, in _insert
dev-celery_worker-1  |     return query.get_compiler(using=using).execute_sql(returning_fields)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1416, in execute_sql
dev-celery_worker-1  |     cursor.execute(sql, params)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 98, in execute
dev-celery_worker-1  |     return super().execute(sql, params)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
dev-celery_worker-1  |     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
dev-celery_worker-1  |     return executor(sql, params, many, context)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
dev-celery_worker-1  |     return self.cursor.execute(sql, params)
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
dev-celery_worker-1  |     raise dj_exc_value.with_traceback(traceback) from exc_value
dev-celery_worker-1  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
dev-celery_worker-1  |     return self.cursor.execute(sql, params)
dev-celery_worker-1  | django.db.utils.IntegrityError: duplicate key value violates unique constraint "core_labelembeddings_label_id_key"
dev-celery_worker-1  | DETAIL:  Key (label_id)=(22) already exists.
```

Reply:
Adding `ignore_conflicts=True` according to the Django docs helped solve this:
[https://docs.djangoproject.com/en/4.0/ref/models/querysets/#django.db.models.query.QuerySet.bulk_create](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#django.db.models.query.QuerySet.bulk_create)

Reply:
Yes this does seem to have fixed things. I also tested updating the label descriptions and the updates correctly impacted the predicted labels.